# NOT DONE

In [5]:
import pandas as pd
import glob, os
import numpy as np
import matplotlib.pyplot as plt

## Source
- [Houston Police Department Crime Statistics](http://www.houstontx.gov/police/cs/crime-stats-archives.htm)
	- years: 2008 - 2017
	- format: Access or Excel

In [6]:
ls crime_data_raw/10years

apr10.xls*  dec10.xls*  jan10.xls*  jun10.xls*  may10.xls*  oct10.xls*
apr11.xls*  dec11.xls*  jan11.xls*  jun11.xls*  may11.xls*  oct11.xls*
apr12.xls*  dec12.xls*  jan12.xls*  jun12.xls*  may12.xls*  oct12.xls*
apr13.xls*  dec13.xls*  jan13.xls*  jun13.xls*  may13.xls*  oct13.xls*
apr14.xls*  dec14.xls*  jan14.xls*  jun14.xls*  may14.xls*  oct14.xls*
apr15.xls*  dec15.xls*  jan15.xls*  jun15.xls*  may15.xls*  oct15.xls*
apr16.xls*  dec16.xls*  jan16.xls*  jun16.xls*  may16.xls*  oct16.xls*
apr17.xls*  dec17.xls*  jan17.xls*  jun17.xls*  may17.xls*  oct17.xls*
aug10.xls*  feb10.xls*  jul10.xls*  mar10.xls*  nov10.xls*  sep10.xls*
aug11.xls*  feb11.xls*  jul11.xls*  mar11.xls*  nov11.xls*  sep11.xls*
aug12.xls*  feb12.xls*  jul12.xls*  mar12.xls*  nov12.xls*  sep12.xls*
aug13.xls*  feb13.xls*  jul13.xls*  mar13.xls*  nov13.xls*  sep13.xls*
aug14.xls*  feb14.xls*  jul14.xls*  mar14.xls*  nov14.xls*  sep14.xls*
aug15.xls*  feb15.xls*  jul15.xls*  mar15.xls*  nov15.xls*  sep15.xls*
aug16.

## combine all files into one dataframe

In [7]:
path = 'crime_data_raw/10years'
all_files = glob.glob(os.path.join(path, "*.xls")) 

df_from_each_file = (pd.read_excel(f) for f in all_files)
df   = pd.concat(df_from_each_file, ignore_index=True)

WARNING *** file size (3078551) not 512 + multiple of sector size (512)
WARNING *** OLE2 inconsistency: SSCS size is 0 but SSAT size is non-zero
WARNING *** file size (2752234) not 512 + multiple of sector size (512)
WARNING *** OLE2 inconsistency: SSCS size is 0 but SSAT size is non-zero


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1006331 entries, 0 to 1006330
Data columns (total 18 columns):
# Of             7613 non-null float64
# Of Offenses    574365 non-null float64
# Offenses       10319 non-null float64
# offenses       334684 non-null float64
Beat             1006314 non-null object
Block Range      671489 non-null object
BlockRange       334684 non-null object
Date             1006326 non-null object
Field11          0 non-null float64
Hour             1006330 non-null object
Offense Type     1006330 non-null object
Offenses         79349 non-null float64
Premise          972734 non-null object
Street Name      671646 non-null object
StreetName       334684 non-null object
Suffix           1006330 non-null object
Type             1006330 non-null object
Unnamed: 1       0 non-null float64
dtypes: float64(7), object(11)
memory usage: 138.2+ MB


In [9]:
df.head(5)

# Of  # Of Offenses  # Offenses  # offenses   Beat  Block Range BlockRange  \
0   NaN            1.0         NaN         NaN  13D20    6600-6699        NaN   
1   NaN            1.0         NaN         NaN  11H40  10100-10199        NaN   
2   NaN            2.0         NaN         NaN  19G20  11400-11499        NaN   
3   NaN            1.0         NaN         NaN  10H60    3700-3799        NaN   
4   NaN            1.0         NaN         NaN  14D30    5100-5199        NaN   

                  Date  Field11 Hour Offense Type  Offenses Premise  \
0  2010-04-17 00:00:00      NaN    0       Murder       NaN     05W   
1  2010-04-08 00:00:00      NaN   20       Murder       NaN     20R   
2  2010-04-01 00:00:00      NaN   22       Murder       NaN     13R   
3  2010-04-17 00:00:00      NaN    1       Murder       NaN     13R   
4  2010-04-08 00:00:00      NaN   23       Murder       NaN     20R   

  Street Name StreetName Suffix Type  Unnamed: 1  
0   HEFFERNAN        NaN      -    -         NaN  
1      LUCORE        NaN      -    -         NaN  
2      CARVEL        NaN      -   LN         NaN  
3     WHEELER        NaN      -    -         NaN  
4  MYRTLEWOOD        NaN      -   DR         NaN

## Lets create a copy

In [12]:
df1 = df.copy()

In [13]:
df1.columns

Index(['# Of', '# Of Offenses', '# Offenses', '# offenses', 'Beat',
       'Block Range', 'BlockRange', 'Date', 'Field11', 'Hour', 'Offense Type',
       'Offenses', 'Premise', 'Street Name', 'StreetName', 'Suffix', 'Type',
       'Unnamed: 1'],
      dtype='object')

In [15]:
df1.head(1)

# Of  # Of Offenses  # Offenses  # offenses   Beat Block Range BlockRange  \
0   NaN            1.0         NaN         NaN  13D20   6600-6699        NaN   

                  Date  Field11 Hour Offense Type  Offenses Premise  \
0  2010-04-17 00:00:00      NaN    0       Murder       NaN     05W   

  Street Name StreetName Suffix Type  Unnamed: 1  
0   HEFFERNAN        NaN      -    -         NaN

## Duplicate columns
We have duplicate columns with diffrent names
- `Block Range` & `BlockRange`
- `StreetName`  & `Street Name`
- `Offenses`  & `# offenses`

lets concat the columns remove nan values and reindex

In [16]:
df1['# offenses'] = pd.concat([df1['# Of'].dropna(),
                               df1['# Of Offenses'].dropna(),
                               df1['# offenses'].dropna(),
                               df1['Offenses'].dropna()]).reindex_like(df1)

df1['BlockRange'] = pd.concat([df1['Block Range'].dropna(),
                               df1['BlockRange'].dropna()]).reindex_like(df1)

df1['StreetName'] = pd.concat([df1['Street Name'].dropna(),
                               df1['StreetName'].dropna()]).reindex_like(df1)

In [18]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1006331 entries, 0 to 1006330
Data columns (total 18 columns):
# Of             7613 non-null float64
# Of Offenses    574365 non-null float64
# Offenses       10319 non-null float64
# offenses       996011 non-null float64
Beat             1006314 non-null object
Block Range      671489 non-null object
BlockRange       1006173 non-null object
Date             1006326 non-null object
Field11          0 non-null float64
Hour             1006330 non-null object
Offense Type     1006330 non-null object
Offenses         79349 non-null float64
Premise          972734 non-null object
Street Name      671646 non-null object
StreetName       1006330 non-null object
Suffix           1006330 non-null object
Type             1006330 non-null object
Unnamed: 1       0 non-null float64
dtypes: float64(7), object(11)
memory usage: 138.2+ MB


## Rename columns

In [ ]:
# df1.rename(columns={
#     '# Of Offenses': '# offenses',
#     'Block Range': 'BlockRange',
#     'Street Name': 'StreetName',}, inplace=True)

In [ ]:
df1.info()

In [ ]:
df1.head(5)

## create a subdataframe with the columns that we want

In [19]:
df1017 = df1[['Date','Beat','BlockRange','StreetName','Offense Type','Premise','# offenses','Hour']]

In [20]:
df1017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1006331 entries, 0 to 1006330
Data columns (total 8 columns):
Date            1006326 non-null object
Beat            1006314 non-null object
BlockRange      1006173 non-null object
StreetName      1006330 non-null object
Offense Type    1006330 non-null object
Premise         972734 non-null object
# offenses      996011 non-null float64
Hour            1006330 non-null object
dtypes: float64(1), object(7)
memory usage: 61.4+ MB


In [21]:
df1017.tail()

Date Beat   BlockRange     StreetName        Offense Type  \
1006326  09/29/2017  UNK  12600-12699    STAFFORD RD             Robbery   
1006327  09/30/2017  UNK  10600-10699       BROOKLET               Theft   
1006328  09/30/2017  UNK  12300-12399  WESTHEIMER RD             Robbery   
1006329  09/30/2017  UNK  13200-13299    BEECHNUT ST  Aggravated Assault   
1006330  11/10/2015  UNK    4900-4999       TEALGATE                Rape   

                                     Premise  # offenses Hour  
1006326            Road, Street, or Sidewalk         1.0    1  
1006327            Road, Street, or Sidewalk         1.0   13  
1006328  Restaurant or Cafeteria Parking Lot         1.0   19  
1006329                   Highway or Freeway         1.0   17  
1006330                   Residence or House         1.0   22

## Now we can inspect df

In [22]:
df1017.Date.unique() ## timestamp string date

array([Timestamp('2010-04-17 00:00:00'), Timestamp('2010-04-08 00:00:00'),
       Timestamp('2010-04-01 00:00:00'), ..., '04/06/2017', '03/24/2017',
       '11/10/2015'], dtype=object)

In [23]:
df1017.Beat.unique()  ##

array(['13D20', '11H40', '19G20', '10H60', '14D30', '17E30', '20G40',
       '18F40', '10H20', '9C20', '11H10', '19G30', '16E40', '19G10',
       '16E20', '19G50', '15E10', '13D40', '14D10', 'UNK', '9C40',
       '10H80', '5F30', '18F50', '7C30', '3B50', '12D10', '20G10',
       '14D20', '17E20', '10H40', '6B60', '3B10', '8C20', '5F20', '7C50',
       '15E30', '16E30', '12D40', '17E40', '18F30', '7C20', '8C60',
       '8C30', '11H50', '20G60', '20G70', '17E10', '3B30', '11H30',
       '2A60', '9C10', '6B10', '6B30', '6B20', '10H50', '12D20', '12D70',
       '6B50', '1A10', '20G30', '6B40', '5F10', '2A20', '13D10', '4F10',
       '9C30', '18F60', '1A30', '19G40', '1A20', '2A10', '7C10', '3B40',
       '7C40', '1A50', '24C30', '14D50', '10H10', '15E20', '10H70',
       '8C10', '20G50', '2A50', '20G80', '15E40', '24C50', '5F40',
       '12D60', '2A30', '4F20', '8C50', '4F30', '18F20', '14D40', '24C10',
       '18F10', '11H20', '10H30', '16E10', '24C40', '12D30', '20G20',
       '12D50', '

In [25]:
df1017.Beat.value_counts(dropna=False)  #

19G10     21243
12D10     21022
1A20      20731
6B60      19887
13D20     19019
3B10      18077
18F20     17395
17E10     17303
5F30      16366
18F30     16327
18F40     16260
1A30      16233
15E40     16216
14D20     15894
20G30     15481
20G50     15380
1A10      15266
2A50      15247
17E40     15116
1A50      14774
6B10      14283
18F50     14171
5F40      13904
7C20      13772
9C40      13253
3B50      12774
6B30      12751
2A30      12669
20G10     12645
19G40     11837
          ...  
'2A40         3
23J20         3
21I40         3
'5F10         3
'24C10        3
'3B30         3
'16E40        3
'4F10         3
21I20         3
'11H50        3
'UNK          3
'11H20        2
23J10         2
'21I10        2
'14D50        2
'12D40        2
'21I50        2
'24C30        2
'24C20        1
'1A40         1
'24C50        1
3B20          1
'24C40        1
'16E30        1
'11H40        1
7C60          1
'8C40         1
'23J50        1
'8C50         1
21I70         1
Name: Beat, Length: 356,

In [26]:
df1017.BlockRange.unique()  #

array(['6600-6699', '10100-10199', '11400-11499', '3700-3799',
       '5100-5199', '11500-11599', '6400-6499', '2300-2399', '7900-7999',
       '8000-8099', '1100-1199', '7300-7399', '6700-6799', '7400-7499',
       '7100-7199', '16300-16399', '15800-15899', '10000-10099',
       '8100-8199', '3100-3199', '10400-10499', '8900-8999', '6300-6399',
       '12600-12699', '1900-1999', '5000-5099', '5800-5899', '5900-5999',
       '2700-2799', '1200-1299', '9800-9899', '4300-4399', '7600-7699',
       '3000-3099', '7200-7299', '1300-1399', '4200-4299', '9000-9099',
       '100-199', '14800-14899', '13300-13399', '0-99', '4800-4899',
       '2200-2299', '11800-11899', '3800-3899', '8300-8399', '7500-7599',
       '12700-12799', '6100-6199', '10200-10299', '14400-14499',
       '8800-8899', '4100-4199', '15500-15599', '9200-9299', '5700-5799',
       '6000-6099', '2000-2099', '6200-6299', '9400-9499', '13200-13299',
       '7800-7899', '5500-5599', '2400-2499', '12800-12899', '1700-1799',
    

In [27]:
df1017.BlockRange.value_counts(dropna=False)  #

900-999          13868
100-199          13831
800-899          11944
700-799          11737
2400-2499        11510
2300-2399        11431
1000-1099        11268
9400-9499        11255
7900-7999        11138
300-399          11007
4400-4499        10895
200-299          10876
500-599          10511
9500-9599        10510
1500-1599        10386
2500-2599        10176
1300-1399        10121
7500-7599        10042
2700-2799         9920
1100-1199         9854
600-699           9645
1400-1499         9608
1200-1299         9558
5000-5099         9527
5800-5899         9339
2100-2199         9114
2000-2099         8992
5900-5999         8981
6000-6099         8962
400-499           8957
                 ...  
28400-28499          1
110600-110699        1
77300-77399          1
35000-35099          1
31300-31399          1
60000-60099          1
175500-175599        1
111100-111199        1
105800-105899        1
31400-31499          1
30500-30599          1
79000-79099          1
42000-42099

In [28]:
df1017.StreetName.unique()  #

array(['HEFFERNAN', 'LUCORE', 'CARVEL', ..., 'GREEN LAKER', 'STAFFORD RD',
       'TEALGATE'], dtype=object)

In [29]:
df1017.StreetName.value_counts(dropna=False)  # name cleanup too!

WESTHEIMER                                                                                                                                                27503
GULF                                                                                                                                                      14140
GESSNER                                                                                                                                                   13904
NORTH                                                                                                                                                     12630
RICHMOND                                                                                                                                                  12376
KATY                                                                                                                                                      10126
SOUTHWEST                               

In [30]:
df1017['Offense Type'].unique()  # good

array(['Murder', 'Rape', 'Robbery', 'Aggravated Assault', 'Burglary',
       'Auto Theft', 'Theft', 1, 'AutoTheft', 'Burglary                 ',
       'Robbery                  ', 'Theft                    ',
       'AutoTheft                ', 'Aggravated Assault       ',
       'Rape                     ', 'Murder                   ', nan],
      dtype=object)

In [31]:
df1017['Offense Type'].value_counts(dropna=False) # good, that its clean not that there were 73591 thefts!! 

Theft                        548360
Burglary                     180870
Auto Theft                    93267
Aggravated Assault            81813
Robbery                       75962
Rape                           7000
AutoTheft                      6649
Theft                          6011
Murder                         1903
Burglary                       1371
Aggravated Assault             1158
AutoTheft                       978
Robbery                         822
Rape                            129
Murder                           23
1                                14
NaN                               1
Name: Offense Type, dtype: int64

In [32]:
df1017.Premise.unique()  # lots of weird values

array(['05W', '20R', '13R', '20A', '18A', '100', '05E', '20D', '18T',
       '250', '040', '13A', '20P', '22M', '140', '070', '14V', '05V',
       '18R', '18C', '18G', '18O', '18M', '18S', '13B', '05D', '22E',
       '240', '18U', '13S', '13H', '210', '120', '080', '24A', '24J',
       '24E', '03B', '05Q', '05B', '05Z', '18N', '05L', '05C', '24F',
       '09D', '24P', '23S', '18P', '18H', '01K', '20G', '05P', '05F',
       '02B', '09P', '05R', '05O', '22P', '11R', '170', '20M', '20L',
       '05M', '18B', '09H', '20V', '22D', '18L', '05X', '13T', '20W',
       '03S', '18D', '20C', '150', '18W', '22H', '20N', '22C', '05Y',
       '20H', '25V', '05A', '190', '24G', '24V', '24C', '05G', '12V',
       '09R', '21V', '060', '19V', '05H', '05U', '11S', '05N', '11V',
       '24S', '11G', '160', '05T', '01T', '01A', '05S', '01B', '22U',
       '24B', '---', '24T', '11L', '02C', '11P', '24M', '01P', '02S',
       '25R', '04V', '11F', '22V', '01R', '11C', 'N', '06', '09V', '23C',
       'Road, St

In [34]:
df1017.Premise.value_counts(dropna=False)  #

20A                                                                                                                                                       74377
20R                                                                                                                                                       74346
18A                                                                                                                                                       65458
Residence or House                                                                                                                                        54178
Apartment Parking Lot                                                                                                                                     44506
13R                                                                                                                                                       41659
Road, Street, or Sidewalk               

In [35]:
df1017['# offenses'].unique()

array([ 1.,  2.,  3.,  4.,  7.,  5.,  6., 12.,  8., 17., 10., 16., 11.,
        9., nan, 33., 13., 29.])

In [37]:
df1017['# offenses'].value_counts(dropna=False)  # 

 1.0     983540
 2.0      10459
NaN       10320
 3.0       1600
 4.0        243
 5.0         79
 6.0         45
 7.0         15
 8.0         10
 9.0          4
 13.0         4
 16.0         3
 10.0         3
 17.0         2
 33.0         1
 29.0         1
 11.0         1
 12.0         1
Name: # offenses, dtype: int64

In [38]:
df1017.Hour.unique()

array([0, 20, 22, 1, 23, 12, 8, 16, 17, 14, 2, 19, 6, 7, 5, 18, 9, 4, 11,
       13, 21, 10, 3, 15, 24, '07', '05', '22', '15', '10', '19', '09',
       '08', '14', "'15", '02', '12', "'07", '16', "'19", "'18", '21',
       '17', '11', '13', '18', '06', "'00", "'05", "'13", "'14", "'22",
       "'23", '04', '03', '23', '20', '00', "'20", "'16", '01', "'17",
       "'06", "'03", "'01", "'09", "'02", "'10", "'11", "'12", "'21",
       "'08", "'04", nan], dtype=object)

In [39]:
df1017.Hour.value_counts(dropna=False)

18     57008
12     56922
17     55531
15     54937
16     53489
19     52654
14     51973
20     51060
13     49339
22     48165
10     47803
11     47177
21     47031
9      44821
8      44571
0      43223
23     39525
7      36164
1      24106
6      23777
2      22343
3      17517
5      15354
4      14223
14       412
15       410
09       405
16       400
11       383
12       377
       ...  
05       129
03       117
01       115
04        94
'00       91
'12       73
'18       69
'22       64
'17       56
'16       53
'20       51
'23       48
'19       48
'21       48
'14       44
'15       43
'08       43
'10       40
'13       38
'11       33
'07       29
'09       25
'02       25
'01       24
'06       17
'05       16
'03       12
'04       10
24         4
NaN        1
Name: Hour, Length: 74, dtype: int64

# Cleanup
## Premise Column

- strip empty spaces :not needed

In [40]:
len(df1017.Premise.unique())

385

In [41]:
df1017['Premise'] = df1017['Premise'].str.strip()

/home/salas/miniconda3/envs/springboard/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [42]:
len(df1017.Premise.unique())

265

In [43]:
df1017.Premise.value_counts(dropna=False)

20A                                          74377
20R                                          74346
18A                                          65458
Residence or House                           55408
Apartment Parking Lot                        45677
13R                                          41659
Road, Street, or Sidewalk                    40121
Apartment                                    39826
20D                                          37014
NaN                                          33597
18O                                          30042
080                                          24960
18R                                          16952
250                                          16634
Driveway                                     15731
Restaurant or Cafeteria Parking Lot          15519
Other Parking Lot                            14724
18P                                          13878
Department or Discount Store                 13194
120                            

# Cleanup
## Offense Type Column

preatty clean!


In [44]:
df1017['Offense Type'].value_counts(dropna=False)

Theft                        548360
Burglary                     180870
Auto Theft                    93267
Aggravated Assault            81813
Robbery                       75962
Rape                           7000
AutoTheft                      6649
Theft                          6011
Murder                         1903
Burglary                       1371
Aggravated Assault             1158
AutoTheft                       978
Robbery                         822
Rape                            129
Murder                           23
1                                14
NaN                               1
Name: Offense Type, dtype: int64

In [45]:
df1017['Offense Type'].unique()

array(['Murder', 'Rape', 'Robbery', 'Aggravated Assault', 'Burglary',
       'Auto Theft', 'Theft', 1, 'AutoTheft', 'Burglary                 ',
       'Robbery                  ', 'Theft                    ',
       'AutoTheft                ', 'Aggravated Assault       ',
       'Rape                     ', 'Murder                   ', nan],
      dtype=object)

In [46]:
len(df1017['Offense Type'].unique())

17

In [47]:
df1017['Offense Type'] = df1017['Offense Type'].str.strip()

/home/salas/miniconda3/envs/springboard/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [48]:
df1017['Offense Type'].value_counts(dropna=False)

Theft                 554371
Burglary              182241
Auto Theft             93267
Aggravated Assault     82971
Robbery                76784
AutoTheft               7627
Rape                    7129
Murder                  1926
NaN                       15
Name: Offense Type, dtype: int64

In [49]:
len(df1017['Offense Type'].unique())

9

# Cleanup
## StreetName Column



In [50]:
df1017.StreetName.value_counts(dropna=False).head()

WESTHEIMER    27503
GULF          14140
GESSNER       13904
NORTH         12630
RICHMOND      12376
Name: StreetName, dtype: int64

In [51]:
len(df1017.StreetName.unique())

27796

In [52]:
df1017['StreetName'] = df1017['StreetName'].str.strip()

/home/salas/miniconda3/envs/springboard/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [53]:
len(df1017.StreetName.unique())

24712

In [54]:
df1017[df1017.StreetName.isnull()].head()

Date  Beat BlockRange StreetName        Offense Type  \
60297  2015-04-10 00:00:00  2A20  3200-3299        NaN  Aggravated Assault   
60389  2015-04-04 00:00:00  2A30        UNK        NaN               Theft   
60404  2015-04-05 00:00:00  2A30        UNK        NaN               Theft   
60438  2015-04-15 00:00:00  2A30        UNK        NaN            Burglary   
70342  2016-04-15 00:00:00  2A30        UNK        NaN               Theft   

                         Premise  # offenses Hour  
60297  Road, Street, or Sidewalk         1.0   12  
60389  Road, Street, or Sidewalk         1.0    0  
60404  Road, Street, or Sidewalk         1.0    7  
60438         Residence or House         1.0   16  
70342          Construction Site         1.0   13

In [55]:
df1017.StreetName.value_counts(dropna=False).head()

WESTHEIMER    27611
GULF          14183
GESSNER       13952
NORTH         12673
RICHMOND      12406
Name: StreetName, dtype: int64

# Cleanup
## BlockRange Column



In [56]:
df1017.BlockRange.value_counts(dropna=False).head()  # 

900-999      13868
100-199      13831
800-899      11944
700-799      11737
2400-2499    11510
Name: BlockRange, dtype: int64

In [57]:
unk = df1017.BlockRange == 'UNK'  # boolean mask

In [62]:
df1017[unk].head()  # 5406

Date   Beat BlockRange StreetName Offense Type  \
53730  2015-04-06 00:00:00  10H10        UNK      CANAL         Rape   
53732  2015-04-01 00:00:00  10H10        UNK     SIDNEY        Theft   
53735  2015-04-18 00:00:00  10H10        UNK      ENNIS         Rape   
53741  2015-04-24 00:00:00  10H10        UNK    MARSDEN     Burglary   
53746  2015-04-23 00:00:00  10H10        UNK     STILES        Theft   

                  Premise  # offenses Hour  
53730  Residence or House         1.0   20  
53732                 NaN         1.0   19  
53735  Residence or House         1.0   10  
53741  Residence or House         1.0   18  
53746   Other Parking Lot         1.0   17

In [66]:
df1017[df1017.BlockRange.isnull()].head()  # 158

Date   Beat BlockRange                 StreetName  \
463032  2014-07-22 00:00:00    NaN        NaN                     WALKER   
463034  2014-07-16 00:00:00    NaN        NaN                      PEASE   
463053  2014-07-27 00:00:00  10H10        NaN                212.5 LENOX   
463599  2014-07-02 00:00:00  10H80        NaN  2600 DUNLAVY + WESTHEIMER   
463917  2014-07-12 00:00:00  12D10        NaN                  EDGEBROOK   

       Offense Type Premise  # offenses Hour  
463032        Theft     NaN         NaN   20  
463034        Theft     NaN         NaN   13  
463053     Burglary     NaN         NaN    7  
463599   Auto Theft     NaN         NaN   12  
463917        Theft     NaN         NaN   11

# Cleanup
## Beat Column



In [67]:
df1017.Beat.unique()

array(['13D20', '11H40', '19G20', '10H60', '14D30', '17E30', '20G40',
       '18F40', '10H20', '9C20', '11H10', '19G30', '16E40', '19G10',
       '16E20', '19G50', '15E10', '13D40', '14D10', 'UNK', '9C40',
       '10H80', '5F30', '18F50', '7C30', '3B50', '12D10', '20G10',
       '14D20', '17E20', '10H40', '6B60', '3B10', '8C20', '5F20', '7C50',
       '15E30', '16E30', '12D40', '17E40', '18F30', '7C20', '8C60',
       '8C30', '11H50', '20G60', '20G70', '17E10', '3B30', '11H30',
       '2A60', '9C10', '6B10', '6B30', '6B20', '10H50', '12D20', '12D70',
       '6B50', '1A10', '20G30', '6B40', '5F10', '2A20', '13D10', '4F10',
       '9C30', '18F60', '1A30', '19G40', '1A20', '2A10', '7C10', '3B40',
       '7C40', '1A50', '24C30', '14D50', '10H10', '15E20', '10H70',
       '8C10', '20G50', '2A50', '20G80', '15E40', '24C50', '5F40',
       '12D60', '2A30', '4F20', '8C50', '4F30', '18F20', '14D40', '24C10',
       '18F10', '11H20', '10H30', '16E10', '24C40', '12D30', '20G20',
       '12D50', '

In [68]:
len(df1017.Beat.unique())

356

In [69]:
df1017['Beat'] = df1017['Beat'].str.strip()

/home/salas/miniconda3/envs/springboard/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [70]:
len(df1017.Beat.unique())

239

In [71]:
df1017[df1017.Beat.isnull()].head()

Date Beat BlockRange StreetName Offense Type Premise  \
463027  2014-07-19 00:00:00  NaN  2100-2199    VINTAGE        Theft     NaN   
463028  2014-07-13 00:00:00  NaN    500-599     W.38TH        Theft     NaN   
463029  2014-07-26 00:00:00  NaN  6100-6199   ATTWATER         Rape     NaN   
463030  2014-07-13 00:00:00  NaN  7800-7899    FM 1960        Theft     NaN   
463031  2014-07-14 00:00:00  NaN  6600-6699      MERRY        Theft     NaN   

        # offenses Hour  
463027         NaN   11  
463028         NaN    8  
463029         NaN   15  
463030         NaN   15  
463031         NaN    9

In [72]:
df1017.Beat.value_counts(dropna=False).head()

19G10    21440
12D10    21252
1A20     20961
6B60     20083
13D20    19181
Name: Beat, dtype: int64

In [73]:
df1017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1006331 entries, 0 to 1006330
Data columns (total 8 columns):
Date            1006326 non-null object
Beat            1006314 non-null object
BlockRange      1006173 non-null object
StreetName      1006231 non-null object
Offense Type    1006316 non-null object
Premise         972734 non-null object
# offenses      996011 non-null float64
Hour            1006330 non-null object
dtypes: float64(1), object(7)
memory usage: 61.4+ MB


## Cleanup

### time column

In [74]:
df1017.Hour.unique()

array([0, 20, 22, 1, 23, 12, 8, 16, 17, 14, 2, 19, 6, 7, 5, 18, 9, 4, 11,
       13, 21, 10, 3, 15, 24, '07', '05', '22', '15', '10', '19', '09',
       '08', '14', "'15", '02', '12', "'07", '16', "'19", "'18", '21',
       '17', '11', '13', '18', '06', "'00", "'05", "'13", "'14", "'22",
       "'23", '04', '03', '23', '20', '00', "'20", "'16", '01', "'17",
       "'06", "'03", "'01", "'09", "'02", "'10", "'11", "'12", "'21",
       "'08", "'04", nan], dtype=object)

In [76]:
df1017.Hour.value_counts(dropna=False).head()  # 

18    57008
12    56922
17    55531
15    54937
16    53489
Name: Hour, dtype: int64

## Cleanup

### Date column
- convert to datetime
- index date colimn
- sort index

In [77]:
df1017.head(5)

Date   Beat   BlockRange  StreetName Offense Type Premise  \
0  2010-04-17 00:00:00  13D20    6600-6699   HEFFERNAN       Murder     05W   
1  2010-04-08 00:00:00  11H40  10100-10199      LUCORE       Murder     20R   
2  2010-04-01 00:00:00  19G20  11400-11499      CARVEL       Murder     13R   
3  2010-04-17 00:00:00  10H60    3700-3799     WHEELER       Murder     13R   
4  2010-04-08 00:00:00  14D30    5100-5199  MYRTLEWOOD       Murder     20R   

   # offenses Hour  
0         1.0    0  
1         1.0   20  
2         2.0   22  
3         1.0    1  
4         1.0   23

In [78]:
df1017['Date'] = pd.to_datetime(df1017['Date'])

df1017 = df1017.set_index('Date').sort_index(ascending=True)

/home/salas/miniconda3/envs/springboard/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [80]:
df1017.head(5)

Beat   BlockRange     StreetName Offense Type  \
Date                                                         
1914-09-08  24C60  12700-12799   LAKE HOUSTON     Burglary   
1914-11-02  18F60    8800-8899       BELLAIRE     Burglary   
1914-12-03  12D20  12800-12899           GULF   Auto Theft   
1915-01-05   3B10    3200-3299  MANGUM RD 180        Theft   
1915-01-14   5F10    7000-7099       WESTVIEW   Auto Theft   

                                          Premise  # offenses Hour  
Date                                                                
1914-09-08                Restaurant or Cafeteria         1.0    7  
1914-11-02  Miscellaneous Business (Non-Specific)         1.0    3  
1914-12-03                                    NaN         1.0   19  
1915-01-05                      Other Parking Lot         1.0   22  
1915-01-14                  Apartment Parking Lot         1.0   23

In [81]:
df1017.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1006331 entries, 1914-09-08 to NaT
Data columns (total 7 columns):
Beat            1006314 non-null object
BlockRange      1006173 non-null object
StreetName      1006231 non-null object
Offense Type    1006316 non-null object
Premise         972734 non-null object
# offenses      996011 non-null float64
Hour            1006330 non-null object
dtypes: float64(1), object(6)
memory usage: 61.4+ MB


## Odd dates
`DatetimeIndex: 1006331 entries, 1914-09-08 to NaT`
- some values are not from this year, lets look

In [87]:
ddf = df1017.loc['2010-01-01':'2017-12-31']  # rows with date from 01,01,10 - 12,31,17
ddf.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1003745 entries, 2010-01-01 to 2017-12-31
Data columns (total 7 columns):
Beat            1003729 non-null object
BlockRange      1003589 non-null object
StreetName      1003646 non-null object
Offense Type    1003731 non-null object
Premise         970213 non-null object
# offenses      993435 non-null float64
Hour            1003745 non-null object
dtypes: float64(1), object(6)
memory usage: 61.3+ MB


In [84]:
dfwrong_date = df1017[:"2009"]  # rows with year 0  upto 2012
dfwrong_date.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2575 entries, 1914-09-08 to 2009-12-31
Data columns (total 7 columns):
Beat            2575 non-null object
BlockRange      2574 non-null object
StreetName      2575 non-null object
Offense Type    2575 non-null object
Premise         2515 non-null object
# offenses      2568 non-null float64
Hour            2575 non-null object
dtypes: float64(1), object(6)
memory usage: 160.9+ KB


In [88]:
ddf.head()

Beat   BlockRange StreetName Offense Type Premise  # offenses  \
Date                                                                         
2010-01-01  14D50    5900-5999   SELINSKY        Theft     18A         1.0   
2010-01-01   9C20      900-999    HOFFMAN      Robbery     20R         1.0   
2010-01-01   9C20    6500-6599  TEXARKANA        Theft     20R         1.0   
2010-01-01  20G30  12500-12599   RICHMOND        Theft     09D         1.0   
2010-01-01   6B30    7200-7299   JADEWOOD         Rape     20R         1.0   

           Hour  
Date             
2010-01-01   20  
2010-01-01    2  
2010-01-01   12  
2010-01-01    7  
2010-01-01   11

# NAN values

- Beat: 16
- BlockRange: 156
- StreetName: 99
- Offense Type: 14
- #offenses : 10310
- Premise: 33532
- Hour: 0

In [90]:
beat_nan = ddf.Beat.isnull()
block_nan = ddf.BlockRange.isnull()
str_nan = ddf.StreetName.isnull()
off_nan = ddf['Offense Type'].isnull()
premise_nan = ddf.Premise.isnull()  #
offnum_nan = ddf['# offenses'].isnull()  #
hour_nan = ddf.Hour.isnull()  #

In [93]:
len(ddf[beat_nan])

16

In [94]:
len(ddf[block_nan])

156

In [95]:
len(ddf[str_nan])

99

In [96]:
len(ddf[off_nan])

14

In [97]:
len(ddf[offnum_nan])

10310

In [98]:
len(ddf[premise_nan])

33532

In [99]:
len(ddf[hour_nan])

0

In [100]:
ddf.head()

Beat   BlockRange StreetName Offense Type Premise  # offenses  \
Date                                                                         
2010-01-01  14D50    5900-5999   SELINSKY        Theft     18A         1.0   
2010-01-01   9C20      900-999    HOFFMAN      Robbery     20R         1.0   
2010-01-01   9C20    6500-6599  TEXARKANA        Theft     20R         1.0   
2010-01-01  20G30  12500-12599   RICHMOND        Theft     09D         1.0   
2010-01-01   6B30    7200-7299   JADEWOOD         Rape     20R         1.0   

           Hour  
Date             
2010-01-01   20  
2010-01-01    2  
2010-01-01   12  
2010-01-01    7  
2010-01-01   11

In [101]:
ddf.Beat.unique()

array(['14D50', '9C20', '20G30', '6B30', '20G10', '17E10', '1A10',
       '16E40', '6B40', '7C30', '10H10', '7C10', '17E30', '2A30', '11H10',
       '2A10', '3B50', '8C60', '18F50', '20G50', '11H30', '5F40', '5F30',
       '8C10', '4F30', '19G10', '20G80', '6B50', '7C40', '10H30', '13D10',
       '7C20', '6B20', '12D70', '6B10', '21I30', '4F10', '5F10', '24C10',
       '16E20', '14D30', '20G40', '3B40', '21I20', '19G50', '17E40',
       '15E30', '12D20', '19G40', '4F20', '18F40', '16E10', '12D10',
       '9C40', '17E20', '14D20', '18F20', '15E10', '16E30', '15E40',
       '3B10', '10H40', '18F60', '8C30', '13D20', '11H20', '24C40',
       '6B60', '1A30', '20G20', '8C20', '5F20', '10H50', '10H60', '19G20',
       '12D30', '2A50', '14D10', '18F30', '8C50', '14D40', '2A60', '3B30',
       '8C40', '10H70', '2A20', '20G60', '13D40', '9C10', '2A40', '18F10',
       '13D30', '1A50', '1A20', '7C50', '11H40', '20G70', '12D50',
       '24C20', '21I50', '9C30', '15E20', '12D60', '12D40', '24C30',

In [102]:
ddf[beat_nan]

Beat BlockRange         StreetName Offense Type Premise  \
Date                                                                 
2013-06-11  NaN  2200-2299          W. DALLAS     Burglary     NaN   
2013-12-01  NaN    700-799         W. DONOVAN        Theft     NaN   
2014-06-05  NaN  2600-2699           LP W 405        Theft     NaN   
2014-06-11  NaN    400-499               MAIN        Theft     NaN   
2014-06-13  NaN  5400-5499    WEST LOOP S SER        Theft     NaN   
2014-06-25  NaN  2100-2199               MAIN        Theft     NaN   
2014-07-03  NaN  8400-8499         S. GESSNER        Theft     NaN   
2014-07-10  NaN    800-899  GESSNER SUITE 970        Theft     NaN   
2014-07-13  NaN  7800-7899            FM 1960        Theft     NaN   
2014-07-13  NaN    500-599             W.38TH        Theft     NaN   
2014-07-14  NaN  6600-6699              MERRY        Theft     NaN   
2014-07-16  NaN        NaN              PEASE        Theft     NaN   
2014-07-19  NaN  2100-2199            VINTAGE        Theft     NaN   
2014-07-22  NaN        NaN             WALKER        Theft     NaN   
2014-07-26  NaN  6100-6199           ATTWATER         Rape     NaN   
2014-07-26  NaN  6100-6199           ATTWATER         Rape     NaN   

            # offenses Hour  
Date                         
2013-06-11         1.0   22  
2013-12-01         1.0   11  
2014-06-05         1.0   13  
2014-06-11         1.0   14  
2014-06-13         1.0   14  
2014-06-25         1.0   14  
2014-07-03         NaN    2  
2014-07-10         NaN   20  
2014-07-13         NaN   15  
2014-07-13         NaN    8  
2014-07-14         NaN    9  
2014-07-16         NaN   13  
2014-07-19         NaN   11  
2014-07-22         NaN   20  
2014-07-26         NaN   14  
2014-07-26         NaN   15